not-script/not-script.c

```diff
@@ -0,0 +1,569 @@
+#define _GNU_SOURCE 1
+#include <inttypes.h>
+#include <limits.h>
+#include <stddef.h>
+#include <assert.h>
+#include <stdbool.h>
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <sys/file.h>
+#include <sys/ioctl.h>
+#include <sys/sysmacros.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+#include <err.h>
+
+#include <linux/major.h>
+#include <linux/loop.h>
+#include <linux/fs.h>
+#include <linux/dm-ioctl.h>
+
+#include <xen/xen.h>
+#include <xen/io/xenbus.h>
+#include <xenstore.h>
+
+static int open_file(const char *path) {
+    return open(path, O_RDONLY | O_NOCTTY | O_CLOEXEC | O_NOFOLLOW);
+}
+
+/* A simple library for loop device ioctls */
+struct loop_context {
+    uint32_t fd;
+};
+
+static int open_loop_dev(uint32_t devnum, bool writable)
+{
+    char buf[sizeof("/dev/loop") + 10];
+    if ((unsigned)snprintf(buf, sizeof buf, "/dev/loop%u", (unsigned)devnum) >= sizeof buf)
+        abort();
+    int flags = (writable ? O_RDWR : O_RDONLY) |
+                   O_EXCL | O_CLOEXEC | O_NOCTTY | O_NOFOLLOW;
+    int res = open(buf, flags);
+    if (res < 0)
+        err(1, "open(\"%s\", %#x)", buf, flags);
+    return res;
+}
+
+static int setup_loop(struct loop_context *ctx,
+                      uint32_t fd,
+                      uint64_t offset,
+                      uint64_t sizelimit,
+                      bool writable, bool autoclear) {
+    int dev_file = -1;
+    int retry_count = 0;
+    const int retry_limit = 5; /* arbitrary */
+#define LOCK_FILE "/run/lock/qubes-script.lock"
+    int lock = open(LOCK_FILE, O_RDWR | O_CREAT | O_NOFOLLOW, 0600);
+    if (lock < 0)
+        err(1, "open(\"%s\")", LOCK_FILE);
+    if (flock(lock, LOCK_EX))
+        err(1, "flock(\"%s\")", LOCK_FILE);
+
+    struct loop_config config = {
+        .fd = fd,
+        .block_size = 0, /* FIXME! */
+        .info = {
+            .lo_offset = offset,
+            .lo_sizelimit = sizelimit,
+            .lo_flags = LO_FLAGS_DIRECT_IO | (autoclear ? LO_FLAGS_AUTOCLEAR : 0) | (writable ? 0 : LO_FLAGS_READ_ONLY),
+        },
+    };
+
+    for (;; retry_count++) {
+        int status = ioctl(ctx->fd, retry_count ? LOOP_CTL_ADD : LOOP_CTL_GET_FREE, -1);
+        if (status < 0)
+            err(1, "ioctl(%d, LOOP_CTL_%s)", ctx->fd, retry_count ? "ADD" : "GET_FREE");
+        char buf[sizeof("/dev/loop") + 10];
```

Comment:
Why not `open_loop_dev()` you used here before? Looks like the change to exit on error made it less useful, so IMO better revert that change (and handle error opening loop-control outside of it), to avoid duplicating this code.